### PR TITLE
make references and lint work in projects outside the gopath

### PR DESCRIPTION
**Pull request by imjustfly**
After go1.11, user's project can be placed outside the gopath directory. Language server doesn't support go1.11 modules currently, this PR makes a little contribution.

This PR makes references and lint work for local packages  and vendored packages. But references still doesn't support third-party packages  only exists in `go.mod` file. Further works needed.

### CODE REVIEW

**Comment by imjustfly** [comment]
Test files need to be modified. Work on it later.

File: langserver/lint.go [hunk 1]
```diff
@@ -54,31 +55,31 @@ func (h *LangHandler) lintPackage(ctx context.Context, bctx *build.Context, conn
...
 			return err
 		}
 
 		for _, f := range p.GoFiles {
 			files = append(files, path.Join(p.Dir, f))
 		}
 	}
-
-	return h.lint(ctx, bctx, conn, []string{path.Join(rootPkg.ImportPath, "/...")}, files)
+	return h.lint(ctx, bctx, conn, []string{path.Join(h.RootFSPath, "/...")}, files)
```
[the remark this answers is not in the document]
Comment (imjustfly) [reply langserver/lint.go]: Um...I don't understand. This works in my situation: 

```
go-langserver|master ⇒
go-langserver|master ⇒ pwd
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver
go-langserver|master ⇒ golint /Users/justfly/golang/src/github.com/sourcegraph/go-langserver/...
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver/langserver/completion.go:17:2: exported var CIKConstantSupported should have comment or be unexported
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver/langserver/fs.go:208:1: exported method HandlerShared.FilePath should have comment or be unexported
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver/langserver/handler_common.go:26:1: exported method HandlerCommon.Reset should have comment or be unexported
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver/langserver/handler_shared.go:106:1: exported method HandlerShared.Reset should have comment or be unexported
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver/langserver/isAlias19.go:15:7: exported const HasAlias should have comment or be unexported
/Users/justfly/golang/src/github.com/sourcegraph/go-langserver/langserver/langserver_test.go:1364:1: context.Context should be the first parameter of a function
```

File: langserver/lint.go [hunk 2]
```diff
@@ -116,7 +116,8 @@ func (l golint) Lint(ctx context.Context, bctx *build.Context, args ...string) (
 	scanner := bufio.NewScanner(stdout)
 
 	for scanner.Scan() {
-		file, line, _, message, err := parseLintResult(scanner.Text())
+		text := scanner.Text()
+		file, line, _, message, err := parseLintResult(text)
```
[the remark this answers is not in the document]
Comment (imjustfly) [reply langserver/lint.go]: this is my fault. I print a log here and forget to place it back.

File: pkg/tools/buildutil.go [hunk 3]
```diff
@@ -19,16 +19,41 @@ import (
...
+
+		wg.Add(1)
+		go func() {
+			allPackages(ctxt, root, dir, ch)
+			wg.Done()
+		}()
+		jobStarted = true
+	}
+
+	if !jobStarted {
```
[the remark this answers is not in the document]
Comment (imjustfly) [reply pkg/tools/buildutil.go]: Yes, this name is much better!

**Comment by imjustfly** [comment]
OK, I will fix these problems and add some tests in next couple of days.

File: langserver/handler_shared.go [hunk 4]
```diff
@@ -32,11 +33,74 @@ func defaultFindPackageFunc(ctx context.Context, bctx *build.Context, importPath
 }
 
 // getFindPackageFunc is a helper which returns h.FindPackage if non-nil, otherwise defaultFindPackageFunc
-func (h *HandlerShared) getFindPackageFunc() FindPackageFunc {
-	if h.FindPackage != nil {
-		return h.FindPackage
+func (h *HandlerShared) getFindPackageFunc(rootPath string) FindPackageFunc {
```
[the remark this answers is not in the document]
Comment (imjustfly) [reply langserver/handler_shared.go]: How to get RootFSPath from HandlerShared without passing it ?

File: langserver/handler_shared.go [hunk 5]
```diff
@@ -32,11 +33,74 @@ func defaultFindPackageFunc(ctx context.Context, bctx *build.Context, importPath
...
+			res        *build.Package
+			err        error
+			importFunc FindPackageFunc
+		)
+
+		if h.FindPackage != nil {
+			importFunc = h.FindPackage
+		} else {
+			importFunc = defaultFindPackageFunc
+		}
```
[the remark this answers is not in the document]
Comment (imjustfly) [reply langserver/handler_shared.go]: Yes, it's not a good solution. 

But supporting project outside GOPATH needs root path. If  we completely use `h.FindPackage` , it's impossible to  support  an outside-gopath-project. 

I think it's better to give type `FindPackageFunc` a new param `rootPath string`. What's your opinion?

File: langserver/lint.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianlopshire) [reply langserver/lint.go]: This seems correct to me. It is also more in line with how other linters expect to receive input.

**Comment by imjustfly** [comment]
Hi @keegancsmith , can you answer my questions in the review conversation ?  I want to fix these problems  ASAP. Thanks.

File: langserver/handler_shared.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (imjustfly) [reply langserver/handler_shared.go]: I added a new param rootPath to type FindPackageFunc.Now, extra logic is moved to defaultFindPackageFunc and getFindPackageFunc is unmodified.

**Pull request merged**